pkg/kvstore/memory.go

```diff
@@ -0,0 +1,342 @@
+func (i *inMemoryModule) setExtraConfig(opts *ExtraOptions) error {
+	return nil
+}
+
+var _ backendModule = &inMemoryModule{}
+
+func newInMemoryClient(db *statedb.DB, clusterName string) *inMemoryClient {
+	table, err := statedb.NewTable(
+		"kvstore-"+clusterName,
+		inMemoryKeyIndex,
+	)
+	if err != nil {
+		panic(err)
+	}
+	if err := db.RegisterTable(table); err != nil {
+		panic(err)
+	}
+	return &inMemoryClient{
+		db:          db,
+		table:       table,
+		clusterName: clusterName,
+	}
+}
+
+type inMemoryObject struct {
+	key   string
+	value []byte
+}
+
+var (
+	inMemoryKeyIndex = statedb.Index[inMemoryObject, string]{
+		Name: "key",
+		FromObject: func(obj inMemoryObject) index.KeySet {
+			return index.NewKeySet(index.String(obj.key))
+		},
+		FromKey:    index.String,
+		FromString: index.FromString,
+		Unique:     true,
+	}
+)
+
+type inMemoryClient struct {
+	db          *statedb.DB
+	table       statedb.RWTable[inMemoryObject]
+	clusterName string
+}
+
+// Close implements BackendOperations.
+func (c *inMemoryClient) Close() {
+}
+
+// Connected implements BackendOperations.
+func (c *inMemoryClient) Connected(ctx context.Context) <-chan error {
+	out := make(chan error)
+	close(out)
+	return out
+}
+
+// CreateOnly implements BackendOperations.
+func (c *inMemoryClient) CreateOnly(ctx context.Context, key string, value []byte, lease bool) (bool, error) {
+	wtxn := c.db.WriteTxn(c.table)
+	defer wtxn.Abort()
+	_, hadOld, _ := c.table.Insert(wtxn, inMemoryObject{
+		key:   key,
+		value: value,
+	})
+	if hadOld {
+		return false, fmt.Errorf("key %q existed", key)
+	}
+	wtxn.Commit()
+	return true, nil
+}
+
+// CreateOnlyIfLocked implements BackendOperations.
+func (c *inMemoryClient) CreateOnlyIfLocked(ctx context.Context, key string, value []byte, lease bool, lock KVLocker) (bool, error) {
+	return c.CreateOnly(ctx, key, value, lease)
+}
+
+// Delete implements BackendOperations.
+func (c *inMemoryClient) Delete(ctx context.Context, key string) error {
+	wtxn := c.db.WriteTxn(c.table)
+	defer wtxn.Abort()
+	_, existed, _ := c.table.Delete(wtxn, inMemoryObject{key: key})
+	if !existed {
+		return fmt.Errorf("key %q did not exist", key)
+	}
+	wtxn.Commit()
+	return nil
+}
+
+// DeleteIfLocked implements BackendOperations.
+func (c *inMemoryClient) DeleteIfLocked(ctx context.Context, key string, lock KVLocker) error {
+	return c.Delete(ctx, key)
+}
+
+// DeletePrefix implements BackendOperations.
+func (c *inMemoryClient) DeletePrefix(ctx context.Context, path string) error {
+	wtxn := c.db.WriteTxn(c.table)
+	defer wtxn.Commit()
+	for obj := range c.table.Prefix(wtxn, inMemoryKeyIndex.Query(path)) {
+		c.table.Delete(wtxn, obj)
+	}
+	return nil
+}
+
+// Get implements BackendOperations.
+func (c *inMemoryClient) Get(ctx context.Context, key string) ([]byte, error) {
+	obj, _, found := c.table.Get(c.db.ReadTxn(), inMemoryKeyIndex.Query(key))
+	if !found {
+		return nil, fmt.Errorf("key %q not found", key)
+	}
```

Comment:
`Get` does not return an error if the key does not exist: https://github.com/joamaki/cilium/blob/pr%2Fjoamaki%2Fcm-service-merger/pkg/kvstore/etcd.go#L1238-L1240